### ISSUE TITLE
[FEATURE] tRNA-seq feature profiling upgrade

### ISSUE THREAD

**Issue by semiller10** [merenlab/anvio#1732]
This is a proposal to improve feature profiling for the identification of tRNAs from tRNA-seq libraries and to benchmark the profiling algorithm with high-scoring tRNA gene predictions from a database.

## Current profiling algorithm
Reads are identified as tRNA by profiling conserved tRNA features, such as the acceptor stem, from the 3' end of the read, or start of the cDNA template. Profiling seeks conserved nucleotides in various features and base pairs in stems. With the default parameterization, a certain number of conserved nucleotides in a feature is allowed to be unconserved, and a certain number of base pairs can be absent from stems. Not every tRNA has the same set of conserved bases, nor are all bases always paired in stems. In tRNA-seq, modification-induced substitutions can convert conserved to unconserved nucleotide reads -- variation in the read is critical in calling modification sites.

The algorithm assigns a tRNA feature profile that minimizes the number of unconserved and unpaired bases in the input sequence. If the algorithm cannot assign a feature, say the D loop, with fewer than the allowed unconserved or unpaired bases for that feature, profiling stops, and a "truncated" tRNA profile is returned -- or, if the profile does not even span the 3' end through the T arm, the rest of the tRNA-seq workflow does not recognize the sequence as tRNA at all. Again, there can be multiple reasons why profiling is stopped by a feature.
- The sequence is not tRNA, e.g., the sequence is rRNA, SRP RNA, or nonsense. True negative.
- The sequence is a chimera, a 3' tRNA fragment joined to another tRNA, rRNA or other RNA at the unprofiled feature. True negative.
- The tRNA feature has more than the allowed number of anomalies. False negative.
- Modification-induced substitutions, possible in conjunction with true nucleotide variation, exceeded the allowed number of anomalies. False negative.
- Modification-induced INDELs messed up the profile at the feature. True negative. Deletions, more common than insertions, are recovered with another algorithm.

Profiles that span beyond the T arm but are then truncated by a bad feature can later be recovered in the tRNA-seq workflow through dereplication of these truncated sequences from fully profiled sequences.

One concern is that common modifications toward the 3' end of the sequence, such as m1A58 in the T loop, can mess up the profile. Another concern is nucleotide switches. G53 is canonically paired with C61 in the T stem, but some tRNAs were found to have these flipped. Tertiary interactions across the tRNA can also be conserved despite the nucleotides involved differing from expectation. Simply reducing feature thresholds, such as the number of unconserved nucleotides allowed in the T loop or T arm, increases the false positive rate of tRNA identifications.

Thresholds and algorithm design should be optimized by systematic testing of a database of high confidence tRNA genes across three domains, such as GtRNAdb (ignoring genes with introns). Testing should include versions of genes with different numbers of random in silico mutations simulating modification-induced substitutions and deletions.

## Proposed algorithm change
Intuitively, profiling should sample beyond features that currently truncate the profile. For example, all of the features 5' of an anomalous T loop may be found to be normal, indicating that the anomalies are natural or are caused by modification-induced substitutions.

Anomalies can only be discounted when there is enough evidence for the rest of the sequence to be tRNA. If an anomaly occurs close to the 5' end, say in the D loop, then there are few remaining features for validation: base pairing in the D stem, U8, and base pairing in the acceptor stem. There should be less tolerance for anything out of the ordinary in these features; perhaps the D loop anomaly can be accommodated if there are no missing base pairs in the D stem or acceptor stem and U8 is present.

Parameterization of features in the current algorithm involves four variables, each of which applies to a subset of features.
- Positions and identities of conserved nucleotides in a feature. This variable is called "Conserved nucleotides" in the feature parameter table. The default value for the anticodon loop is "1,T;5,R", indicating position 33, the second position in the loop from the 5' end, is typically a T, and position 37, the sixth position in the loop, is typically a purine.
- "Number allowed unconserved". The default value is "1" for the anticodon loop -- one of the conserved positions 33 or 37 is allowed to deviate.
- "Number allowed unpaired", by default, "1" for any of the four stems.
- "Allowed lengths", by default, "1-3" nucleotides for the alpha and beta segments of the D loop and "4-5,9-23" for the V loop.

The algorithm can be upgraded to allow profiling to continue after a feature has exceeded the numbers of nucleotides allowed unconserved and unpaired. Every feature will have a new parameter, "Number allowed further anomalies." A value of "0" prevents further profiling. A value of "1" indicates that one unconserved nucleotide or absent base pair is allowed in the more 5' features. Appropriate values might be 3 for the T loop and 0 for the D loop.

The triggering feature should contribute to the "further anomalies" when it exceeds the unconserved and unpaired thresholds by more than 1. For instance, if the anticodon stem has 3 missing base pairs, then given the default threshold of 1, it would contribute 1 "further anomaly." An appropriate number of allowed further anomalies might be 1 for the anticodon stem, meaning that the profile would stop at that feature rather than continuing in this example.

It should be relatively straightforward to extend the profiling algorithm in `trnaidentifier` with these changes.

**Comment by meren** [comment]
Thank you very much for the very clear description, Sam. This makes me wonder whether we were missing a lot of tRNAs during profiling that contained a lot of modifications. I am wondering whether you realized this after seeing something (i.e., excluded tRNAs that looked like false negatives) or concluded that this was necessary after a random epiphany :)

My intuition tells me that this really really calls for a statistical thinking rather than hard-coded cutoffs, but even if we can quantifying the distribution of features that stray away from the expected tRNA structure perhaps a good model would present itself automatically.

Thank you for your meticulous work and thoughts on these critical aspects of the profiling.

**Comment by semiller10** [comment]
Observations and introspection led me to this conclusion. One observation was the relative frequency of nucleotide switching that preserves base pairing (either canonical or non-canonical, often long-rang across the secondary structure). Sometimes this affects conserved nucleotides. Another observation was that modifications in the T loop, especially m1A58, seem to be less common than I expected for certain bacterial groups' isoacceptors that usually have them. This suggests that tRNAs lacking that modification are preferentially profiled, allowing long tRNA sequences to still be profiled and modifications closer to the 5' end, like m122A, to still be found. Again, when a profile is messed up before the end of the T arm, a short way into the profile, the sequence is called non-tRNA and effectively discarded rather than tRNA with a truncated profile and potentially salvaged later.

Given the superior results of profiling compared to Bowtie mapping, in both read recruitment and the identification of 5' modifications, and the inherent biases of that approach, improvement in the false negative rate and reduction of remaining biases in profiling is icing on the cake. Benchmarking against a reference database will be very informative and add to the persuasive power of the paper. I think there will be a general optimal parameterization that accommodates the vast majority of tRNAs. To me, that provides the statistical argument. I can't grasp how the algorithm could include a statistical dimension. For example, you could argue that different groups have different statistical properties of their sequences, so the algorithm should vary by taxonomy, but the algorithm has no idea of taxonomy.

**Comment by meren** [comment]
> I think there will be a general optimal parameterization that accommodates the vast majority of tRNAs. To me, that provides the statistical argument.

Fair enough. I was thinking that perhaps an approach that gives a likelihood of a sequence being a tRNA despite deviations from expectations would be a better one than one that has a strict cutoff for binary decision, but it may not even be computationally feasible.

**Comment by semiller10** [comment]
Ah, I see. The benchmarking would certainly generate statistics like that, i.e., a gene that the algorithm identifies as tRNA with 0 anomalies has an X% likelihood of being correct (this could be done for all tRNAs or for subsets of isoacceptors and taxonomic groups), and a gene that the algorithm identifies as tRNA with 1 anomalies has a Y% likelihood, which will probably be lower than X. That could conceivably be reported by the algorithm, too.

**Comment by semiller10** [comment]
I'm not sure how much I would trust predictions for tRNA-seq reads with modifications given training data with in silico substitutions and deletions, though.
